pkg/deployments/tasks/20230215-single-recipient-gauge-factory-v2/test/task.fork.ts

```diff
@@ -123,168 +124,207 @@ describeForkTest('SingleRecipientGaugeFactory V2', 'mainnet', 16627100, function
     await BALTokenAdmin.update_mining_parameters();
   });
 
-  it('create gauge', async () => {
-    // We use an EOA as the single recipient; in practice it will probably be a contract.
-    const tx = await factory.create(recipient.address, weightCap);
-    const event = expectEvent.inReceipt(await tx.wait(), 'GaugeCreated');
+  enum RecipientMode {
+    BasicRecipient,
+    FeeDistributorRecipient,
+  }
 
-    gauge = await task.instanceAt('SingleRecipientGauge', event.args.gauge);
-    expect(await gauge.getRecipient()).to.equal(recipient.address);
-
-    expect(await factory.isGaugeFromFactory(gauge.address)).to.be.true;
+  context('with a basic recipient', () => {
+    itWorksLikeACappedSingleRecipientGauge(RecipientMode.BasicRecipient);
   });
 
-  it('grant permissions', async () => {
-    // We are not using the GaugeAdder at the moment, so gauges shall be added to the gauge controller directly.
-    // Therefore, we just grant the admin the permission to add the gauge to the controller, and perform a checkpoint.
-    const govMultisig = await impersonate(GOV_MULTISIG);
-    const gaugeControllerTask = new Task('20220325-gauge-controller', TaskMode.READ_ONLY, getForkedNetwork(hre));
-    await authorizer
-      .connect(govMultisig)
-      .grantRole(gaugeControllerTask.actionId('GaugeController', 'add_gauge(address,int128)'), admin.address);
-
-    await authorizer
-      .connect(govMultisig)
-      .grantRole(await authorizerAdaptor.getActionId(gauge.interface.getSighash('checkpoint')), admin.address);
+  context('with a FeeDistributor recipient', () => {
+    itWorksLikeACappedSingleRecipientGauge(RecipientMode.FeeDistributorRecipient);
   });
 
-  it('add gauge to gauge controller directly via AuthorizerAdaptor', async () => {
-    // Using 2 as Ethereum gauge type, but it could be any of the existing types since they all have the same
-    // relative weight in the controller.
-    const calldata = gaugeController.interface.encodeFunctionData('add_gauge(address,int128)', [gauge.address, 2]);
-    await authorizerAdaptor.connect(admin).performAction(gaugeController.address, calldata);
-
-    expect(await gaugeController.gauge_exists(gauge.address)).to.be.true;
-  });
+  function itWorksLikeACappedSingleRecipientGauge(mode: RecipientMode) {
+    // We're going to create gauges, vote for them, have time pass, etc. Because of that, we take a snapshot before we
+    // do any of that, and restore it at the end.
+    let snapshot: SnapshotRestorer;
```

Comment:
uhmm why not just use an empty `sharedBeforeEach`?

Reply:
Because that'd cause for each `it` to revert to the setup snapshot, which we don't want (since each `it` performs steps that we build on top of).